File: lib/karabinex/validator.ex
defmodule Karabinex.Validator do
  alias Karabinex.{Key, Command}

  @valid_kinds [:app, :quit, :kill, :sh, :remap, :raycast]
  @valid_opts [:repeat]

  @spec validate!(map()) :: map() | no_return()
  def validate!(defs) do
    validate_definitions!(defs, _depth = 0)
    defs
  end

  @spec validate_definitions!(map(), non_neg_integer()) :: :ok | no_return()
  defp validate_definitions!(defs, depth) do
    validate_not_empty!(defs, depth)
    validate_hook!(defs, depth)
    validate_no_duplicates!(defs)
    Enum.each(defs, &validate_definition!(&1, depth))
  end

  @spec validate_hook!(map(), non_neg_integer()) :: :ok | no_return()
  defp validate_hook!(%{__hook__: _}, 0) do
    raise "__hook__ cannot be used at top level"
  end

  defp validate_hook!(%{__hook__: {_kind, _arg, opts}}, _depth) do
    raise "Hook cannot have options: #{inspect(opts)}"
  end

  defp validate_hook!(%{__hook__: {kind, arg}}, _depth) do
    validate_kind!(kind)
    validate_arg!(arg)
  end

  defp validate_hook!(%{}, _depth), do: :ok

  @spec validate_not_empty!(map(), non_neg_integer()) :: :ok | no_return()
  defp validate_not_empty!(%{} = defs, 0) when map_size(defs) == 0 do
    raise "Config cannot be empty"
  end

  defp validate_not_empty!(%{} = defs, _depth) when map_size(defs) == 0 do
    raise "Empty keymap is not allowed"
  end

  defp validate_not_empty!(_defs, _depth), do: :ok

  @spec validate_no_duplicates!(map()) :: :ok | no_return()
  defp validate_no_duplicates!(defs) do
    defs
    |> find_duplicate_keys()
    |> then(fn
      [] ->
        :ok

      dups ->
        keys =
          dups
          |> Enum.map(fn keys -> Enum.map_join(keys, ", ", &inspect/1) end)
          |> Enum.join("; ")

        raise("Duplicate keys detected: #{keys}")
    end)
  end

  @spec find_duplicate_keys(map()) :: [[atom()]]
  defp find_duplicate_keys(defs) do
    defs
    |> Enum.reject(fn {key, _} -> key == :__hook__ end)
    |> Enum.map(fn {key, _} -> {key, Key.new(key)} end)
    |> Enum.group_by(fn {_raw, parsed} -> {parsed.code, parsed.modifiers} end)
    |> Enum.filter(fn {_normalized, entries} -> length(entries) > 1 end)
    |> Enum.map(fn {_normalized, entries} -> Enum.map(entries, &elem(&1, 0)) end)
  end

  @spec validate_definition!({atom(), term()}, non_neg_integer()) :: :ok | no_return()
  defp validate_definition!({:__hook__, _value}, _depth) do
    :ok
  end

  defp validate_definition!({key, %{} = nested}, depth) do
    validate_key!(key)
    validate_definitions!(nested, depth + 1)
  end

  defp validate_definition!({key, {kind, arg}}, depth) do
    validate_key!(key)
    validate_kind!(kind)
    validate_arg!(arg)
    validate_no_repeat_at_top!([], depth, key)
  end

  defp validate_definition!({key, {kind, arg, opts}}, depth) do
    validate_key!(key)
    validate_kind!(kind)
    validate_arg!(arg)
    validate_opts!(opts)
    validate_no_repeat_at_top!(opts, depth, key)
  end

  @spec validate_key!(atom()) :: :ok | no_return()
  defp validate_key!(key) do
    _ = Key.new(key)
    :ok
  rescue
    e -> reraise "Invalid key #{inspect(key)}: #{Exception.message(e)}", __STACKTRACE__
  end

  @spec validate_kind!(Command.kind()) :: :ok
  defp validate_kind!(kind) when kind in @valid_kinds, do: :ok

  @spec validate_kind!(atom()) :: no_return()
  defp validate_kind!(kind) do
    raise "Unknown command type: #{inspect(kind)}. Valid types: #{inspect(@valid_kinds)}"
  end

  @spec validate_arg!(binary()) :: :ok
  defp validate_arg!(arg) when is_binary(arg), do: :ok

  @spec validate_arg!(term()) :: no_return()
  defp validate_arg!(arg) do
    raise "Command argument must be a string, got: #{inspect(arg)}"
  end

  @spec validate_opts!(keyword()) :: :ok | no_return()
  defp validate_opts!(opts) do
    Enum.each(opts, fn {key, _value} ->
      if key not in @valid_opts do
        raise "Unknown option: #{inspect(key)}. Valid options: #{inspect(@valid_opts)}"
      end
    end)
  end

  @spec validate_no_repeat_at_top!(keyword(), non_neg_integer(), atom()) :: :ok | no_return()
  defp validate_no_repeat_at_top!(opts, 0, key) do
    case opts[:repeat] do
      nil ->
        :ok

      value ->
        raise "repeat: #{inspect(value)} cannot be used at top level (key: #{inspect(key)})"
    end
  end

  defp validate_no_repeat_at_top!(_opts, _depth, _key), do: :ok
end

File: src/validator.ts
import {
  type CommandDef,
  type CommandKind,
  commandKinds,
  isCommandKind,
} from "./command";
import type { Binding } from "./jsonConfig";
import { KeymapDef } from "./jsonConfig";
import { Key } from "./key";
import type { KeyCodes } from "./keyCodes";

export function validate(defs: KeymapDef, keyCodes: KeyCodes): void {
  validateDefinitions(defs, 0, keyCodes);
}

function validateDefinitions(
  defs: KeymapDef,
  depth: number,
  keyCodes: KeyCodes,
): void {
  validateNotEmpty(defs, depth);
  validateHook(defs, depth);
  validateNoDuplicates(defs, keyCodes);

  for (const [key, value] of defs.entries) {
    validateDefinition(key, value, depth, keyCodes);
  }
}

function validateHook(defs: KeymapDef, depth: number): void {
  if (!defs.hook) {
    return;
  }

  if (depth === 0) {
    throw new Error("__hook__ cannot be used at top level");
  }

  if (defs.hook.repeat !== undefined) {
    throw new Error(
      `Hook cannot have options: ${JSON.stringify({ repeat: defs.hook.repeat })}`,
    );
  }

  validateKind(defs.hook.kind);
  validateArg(defs.hook.arg);
}

function validateNotEmpty(defs: KeymapDef, depth: number): void {
  if (defs.entries.size === 0 && depth === 0) {
    throw new Error("Config cannot be empty");
  }

  if (defs.entries.size === 0 && depth > 0) {
    throw new Error("Empty keymap is not allowed");
  }
}

function validateNoDuplicates(defs: KeymapDef, keyCodes: KeyCodes): void {
  const groups = new Map<string, string[]>();

  for (const [key] of defs.entries) {
    const parsed = Key.parse(key, keyCodes);
    const groupKey = `${parsed.code.type}:${parsed.code.code}:${parsed.modifierSetKey()}`;
    const list = groups.get(groupKey);
    if (list) {
      list.push(key);
    } else {
      groups.set(groupKey, [key]);
    }
  }

  const duplicates = [...groups.values()].filter(
    (entries) => entries.length > 1,
  );
  if (duplicates.length === 0) {
    return;
  }

  const keys = duplicates
    .map((entries) => entries.map((entry) => JSON.stringify(entry)).join(", "))
    .join("; ");
  throw new Error(`Duplicate keys detected: ${keys}`);
}

function validateDefinition(
  key: string,
  value: Binding,
  depth: number,
  keyCodes: KeyCodes,
): void {
  if (value instanceof KeymapDef) {
    validateTopLevelModifiers(key, depth, keyCodes);
    validateKey(key, keyCodes);
    validateDefinitions(value, depth + 1, keyCodes);
    return;
  }

  validateTopLevelModifiers(key, depth, keyCodes);
  validateKey(key, keyCodes);
  validateKind(value.kind);
  validateArg(value.arg);
  validateNoRepeatAtTop(value, depth, key);
}

function validateKey(key: string, keyCodes: KeyCodes): void {
  try {
    Key.parse(key, keyCodes);
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    throw new Error(`Invalid key ${JSON.stringify(key)}: ${message}`);
  }
}

function validateKind(kind: CommandKind): void {
  if (!isCommandKind(kind)) {
    throw new Error(
      `Unknown command type: ${JSON.stringify(kind)}. Valid types: ${JSON.stringify(commandKinds)}`,
    );
  }
}

function validateArg(arg: unknown): void {
  if (typeof arg !== "string") {
    throw new Error(
      `Command argument must be a string, got: ${JSON.stringify(arg)}`,
    );
  }
}

function validateNoRepeatAtTop(
  command: CommandDef,
  depth: number,
  key: string,
): void {
  if (depth === 0 && command.repeat !== undefined) {
    throw new Error(
      `repeat: ${JSON.stringify(command.repeat)} cannot be used at top level (key: ${JSON.stringify(key)})`,
    );
  }
}

function validateTopLevelModifiers(
  key: string,
  depth: number,
  keyCodes: KeyCodes,
): void {
  if (depth !== 0) {
    return;
  }

  const parsed = Key.parse(key, keyCodes);
  if (!parsed.hasModifiers()) {
    throw new Error(
      `Top-level key must include modifiers: ${JSON.stringify(key)}`,
    );
  }
}
